src/algorithms/tonal/nnlschroma.cpp

```diff
@@ -0,0 +1,553 @@
+    0.942058, 0.958281, 0.971942, 0.982963, 0.991283, 0.996856, 0.999650,
+    0.999650, 0.996856, 0.991283, 0.982963, 0.971942, 0.958281, 0.942058,
+    0.923362, 0.902299, 0.878986, 0.853553, 0.826144, 0.796910, 0.766016,
+    0.733634, 0.699946, 0.665140, 0.629410, 0.592956, 0.555982, 0.518696,
+    0.481304, 0.444018, 0.407044, 0.370590, 0.334860, 0.300054, 0.266366,
+    0.233984, 0.203090, 0.173856, 0.146447, 0.121014, 0.097701, 0.076638,
+    0.057942, 0.041719, 0.028058, 0.017037, 0.008717, 0.003144, 0.000350};
+
+void NNLSChroma::configure() {
+  _frameSize = parameter("frameSize").toInt();
+  _sampleRate = parameter("sampleRate").toReal();
+  _whitening = parameter("spectralWhitening").toReal();
+  _spectralShape = parameter("spectralShape").toReal();
+  _useNNLS = parameter("useNNLS").toBool();
+
+  string tuningString = parameter("tuningMode").toString();
+  if (tuningString == "local")
+    _tuningMode = true;
+  if (tuningString == "global")
+    _tuningMode = false;
+
+  string normalizationString = parameter("chromaNormalisation").toString();
+  if (normalizationString == "none")
+    _doNormalizeChroma = 0;
+  if (normalizationString == "maximum")
+    _doNormalizeChroma = 1;
+  if (normalizationString == "L1")
+    _doNormalizeChroma = 2;
+  if (normalizationString == "L2")
+    _doNormalizeChroma = 3;
+
+
+  // Make things for tuning estimation.
+  _sinvalues.clear();
+  _cosvalues.clear();
+	for (int iBPS = 0; iBPS < nBPS; ++iBPS) {
+    _sinvalues.push_back(sin(2 * M_PI * (iBPS * 1.0 / nBPS)));
+    _cosvalues.push_back(cos(2 * M_PI * (iBPS * 1.0 / nBPS)));
+  }
+
+	// Make hamming window of length 1/2 octave.
+	int hamwinlength = nBPS * 6 + 1;
+
+  Real hamwinsum = 0;
+  _hw.clear();
+  for (int i = 0; i < hamwinlength; ++i) { 
+    _hw.push_back(0.54 - 0.46 * cos((2 * M_PI * i) / (hamwinlength - 1)));    
+    hamwinsum += 0.54 - 0.46 * cos((2 * M_PI * i) / (hamwinlength - 1));
+  }
+
+  for (int i = 0; i < hamwinlength; ++i) _hw[i] = _hw[i] / hamwinsum;
+
+  int tempn = nNote * _frameSize;
+
+  vector<Real> tempkernel(tempn);
+
+  logFreqMatrix(_sampleRate, _frameSize, tempkernel);
+  _kernelValue.clear();
+  _kernelFftIndex.clear();
+  _kernelNoteIndex.clear();
+  int countNonzero = 0;
+  for (int iNote = 0; iNote < nNote; ++iNote) {
+    for (int iFFT = 0; iFFT <static_cast<int>(_frameSize); ++iFFT) {
+      if (tempkernel[iFFT + _frameSize * iNote] > 0) {
+        _kernelValue.push_back(tempkernel[iFFT + _frameSize * iNote]);
+        if (tempkernel[iFFT + _frameSize * iNote] > 0) {
+            countNonzero++;
+        }
+        _kernelFftIndex.push_back(iFFT);
+        _kernelNoteIndex.push_back(iNote);				
+      }
+    }
+  }
+
+  _dict.assign(nNote * 84, 0.f);
+  for (int i = 0; i < nNote * 84; ++i) _dict[i] = 0.0;
+
+  dictionaryMatrix(_dict, _spectralShape);
+}
+
+void NNLSChroma::reset() {
+  configure();
+}
+
+void NNLSChroma::compute() {
```

Comment:
It may be useful to add code comments that mark which parts of the code were taken in from 3rd parties without a proper refactoring.